<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Estimate diversity of repertoire — repDiversity • immunarch</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Estimate diversity of repertoire — repDiversity" />
<meta property="og:description" content="This is a utility function to estimate the diversity of species or objects
in the given distribution.
Note: functions will check if .data is a distribution of a random variable (sum == 1) or not.
To force normalisation and / or to prevent this, set .do.norm to TRUE (do normalisation)
or FALSE (don't do normalisation), respectively." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-128418614-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-128418614-2');
</script>


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunarch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/v1_introduction.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/v1_introduction.html">Introduction to immunarch</a>
    </li>
    <li>
      <a href="../articles/v2_data.html">Working with data in immunarch</a>
    </li>
    <li>
      <a href="../articles/v3_basic_analysis.html">Basic analysis and clonality</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Explore and compare repertoires</li>
    <li>
      <a href="../articles/web_only/v4_overlap.html">Repertoire overlap and public clonotypes</a>
    </li>
    <li>
      <a href="../articles/web_only/v5_gene_usage.html">Gene usage analysis</a>
    </li>
    <li>
      <a href="../articles/web_only/v6_diversity.html">Diversity estimation</a>
    </li>
    <li>
      <a href="../articles/web_only/v11_db.html">Annotate clonotypes using immune receptor databases</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preparing to publication</li>
    <li>
      <a href="../articles/web_only/v7_fixvis.html">Make your plots publication-ready with fixVis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Advanced methods</li>
    <li>
      <a href="../articles/web_only/v8_tracking.html">Tracking clonotypes across time points</a>
    </li>
    <li>
      <a href="../articles/web_only/v9_kmers.html">Kmer and sequence motif analysis and visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch/releases">Changelog</a>
</li>
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://immunomind.io">ImmunoMind</a>
</li>
<li>
  <a href="http://twitter.com/immunomind">
    <span class="fa fa-lg fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/immunomind/immunarch">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate diversity of repertoire</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunomind/immunarch/blob/master/R/diversity.R'><code>R/diversity.R</code></a></small>
    <div class="hidden name"><code>repDiversity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is a utility function to estimate the diversity of species or objects
in the given distribution.</p>
<p>Note: functions will check if .data is a distribution of a random variable (sum == 1) or not.
To force normalisation and / or to prevent this, set .do.norm to TRUE (do normalisation)
or FALSE (don't do normalisation), respectively.</p>
    </div>

    <pre class="usage"><span class='fu'>repDiversity</span>(
  <span class='no'>.data</span>,
  <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>"chao1"</span>,
  <span class='kw'>.col</span> <span class='kw'>=</span> <span class='st'>"aa"</span>,
  <span class='kw'>.max.q</span> <span class='kw'>=</span> <span class='fl'>6</span>,
  <span class='kw'>.min.q</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>.q</span> <span class='kw'>=</span> <span class='fl'>5</span>,
  <span class='kw'>.step</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.quantile</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.025</span>, <span class='fl'>0.975</span>),
  <span class='kw'>.extrapolation</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.perc</span> <span class='kw'>=</span> <span class='fl'>50</span>,
  <span class='kw'>.norm</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>.verbose</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>,
  <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>The data to be processed. Can be <a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a>,
data.table, or a list of these objects.</p>
<p>Every object must have columns in the immunarch compatible format.
<a href='immunr_data_format.html'>immunarch_data_format</a></p>
<p>Competent users may provide advanced data representations:
DBI database connections, Apache Spark DataFrame from copy_to or a list
of these objects. They are supported with the same limitations as basic objects.</p>
<p>Note: each connection must represent a separate repertoire.</p></td>
    </tr>
    <tr>
      <th>.method</th>
      <td><p>Pick a method used for estimation out of a following list: chao1,
hill, div, gini.simp, inv.simp, gini, raref, d50, dxx.</p></td>
    </tr>
    <tr>
      <th>.col</th>
      <td><p>A string that specifies the column(s) to be processed. Pass one of the
following strings, separated by the plus sign: "nt" for nucleotide sequences,
"aa" for amino acid sequences, "v" for V gene segments, "j" for J gene segments. E.g.,
pass "aa+v" to compute diversity estimations on CDR3 amino acid sequences paired with V gene segments, i.e.,
in this case a unique clonotype is a pair of CDR3 amino acid and V gene segment.
Clonal counts of equal clonotypes will be summed up.</p></td>
    </tr>
    <tr>
      <th>.max.q</th>
      <td><p>The max hill number to calculate (default: 5).</p></td>
    </tr>
    <tr>
      <th>.min.q</th>
      <td><p>Function calculates several hill numbers. Set the min (default: 1).</p></td>
    </tr>
    <tr>
      <th>.q</th>
      <td><p>q-parameter for the Diversity index.</p></td>
    </tr>
    <tr>
      <th>.step</th>
      <td><p>Rarefaction step's size.</p></td>
    </tr>
    <tr>
      <th>.quantile</th>
      <td><p>Numeric vector with quantiles for confidence intervals.</p></td>
    </tr>
    <tr>
      <th>.extrapolation</th>
      <td><p>An integer. An upper limit for the number of clones to extrapolate to.
Pass 0 (zero) to turn extrapolation subroutines off.</p></td>
    </tr>
    <tr>
      <th>.perc</th>
      <td><p>Set the percent to dXX index measurement.</p></td>
    </tr>
    <tr>
      <th>.norm</th>
      <td><p>Normalise rarefaction curves.</p></td>
    </tr>
    <tr>
      <th>.verbose</th>
      <td><p>If T then output progress.</p></td>
    </tr>
    <tr>
      <th>.do.norm</th>
      <td><p>One of the three values - NA, T or F. If NA then check for distrubution (sum(.data) == 1)
and normalise if needed with the given laplace correction value. if T then do normalisation and laplace
correction. If F then don't do normalisaton and laplace correction.</p></td>
    </tr>
    <tr>
      <th>.laplace</th>
      <td><p>A numeric value, which is used as a pseudocount for Laplace
smoothing.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>div, gini, gini.simp, inv.simp, raref return numeric vector of length 1
with value.</p>
<p>chao1 returns 4 values: estimated number of species, standart deviation of
this number and two 95<!-- % confidence intervals for the species number. --></p>
<p>hill returns a vector of specified length <code>.max.q - .min.q</code></p>
<p>If input data is a single immune repertoire, then the function returns a numeric vector
with diversity statistics.</p>
<p>Otherwise, it returns a numeric matrix with diversity statistics for all input repertoires.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>- True diversity, or the effective number of types, refers to the number
of equally-abundant types needed for the average proportional abundance
of the types to equal that observed in the dataset of interest
where all types may not be equally abundant.</p>
<p>- Inverse Simpson index is the effective number of types that is obtained when
the weighted arithmetic mean is used to quantify average
proportional abundance of types in the dataset of interest.</p>
<p>- The Gini coefficient measures the inequality among values
of a frequency distribution (for example levels of income). A Gini coefficient of zero
expresses perfect equality, where all values are the same (for example, where everyone
has the same income). A Gini coefficient of one (or 100 percents ) expresses maximal inequality
among values (for example where only one person has all the income).</p>
<p>- The Gini-Simpson index is the probability of interspecific encounter, i.e., probability that two entities
represent different types.</p>
<p>- Chao1 estimator is a nonparameteric asymptotic estimator of species richness (number of species in a population).</p>
<p>- Rarefaction is a technique to assess species richness from the results of sampling through extrapolation.</p>
<p>- Hill numbers are a mathematically  unified family of  diversity indices (differing among themselves
only by an exponent q).</p>
<p>- d50 is a recently developed immune diversity estimate. It calculates the minimum number of distinct clonotypes
amounting to greater than or equal to 50 percent of a total of sequencing reads obtained
following amplification and sequencing</p>
<p>- dXX is a similar to d50 index where XX corresponds to desirable percent of total sequencing reads.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='repOverlap.html'>repOverlap</a>, <a href='entropy.html'>entropy</a>, <a href='repClonality.html'>repClonality</a>
Rarefaction wiki
<a href='https://en.wikipedia.org/wiki/Rarefaction_(ecology)'>https://en.wikipedia.org/wiki/Rarefaction_(ecology)</a>
Hill numbers paper
<a href='https://www.uvm.edu/~ngotelli/manuscriptpdfs/ChaoHill.pdf'>https://www.uvm.edu/~ngotelli/manuscriptpdfs/ChaoHill.pdf</a>
Diversity wiki
<a href='https://en.wikipedia.org/wiki/Measurement_of_biodiversity'>https://en.wikipedia.org/wiki/Measurement_of_biodiversity</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>immdata</span>)

<span class='co'># chao1</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'chao1'</span>)

<span class='co'># Hill numbers</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'hill'</span>, <span class='kw'>.max.q</span> <span class='kw'>=</span> <span class='fl'>6</span>,
             <span class='kw'>.min.q</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>)

<span class='co'># diversity</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'dev'</span>, <span class='kw'>.q</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>)

<span class='co'># Gini-Simpson</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'gini.simp'</span>, <span class='kw'>.q</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>)

<span class='co'># inverse Simpson</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'inv.simp'</span>, <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>)

<span class='co'># Gini coefficient</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'gini.coef'</span>, <span class='kw'>.do.norm</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.laplace</span> <span class='kw'>=</span> <span class='fl'>0</span>)

<span class='co'># rarefaction</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'raref'</span>, <span class='kw'>.step</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>.quantile</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>.025</span>, <span class='fl'>.975</span>),
<span class='kw'>.extrapolation</span> <span class='kw'>=</span> <span class='fl'>200000</span>, <span class='kw'>.verbose</span> <span class='kw'>=</span> <span class='no'>T</span>)

<span class='co'># d50</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'d50'</span>)

<span class='co'># dXX</span>
<span class='fu'>repDiversity</span>(<span class='kw'>.data</span> <span class='kw'>=</span> <span class='no'>immdata</span>, <span class='kw'>.method</span> <span class='kw'>=</span> <span class='st'>'dXX'</span>, <span class='kw'>.perc</span> <span class='kw'>=</span> <span class='fl'>10</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vadim I. Nazarov, Vasily O. Tsvetkov, Eugene Rumynskiy, Anna Lorenc, Daniel J. Moore, Victor Greiff, <a href='https://immunomind.io'>ImmunoMind</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '94207562f34b455c0796790d104e3549',
    indexName: 'immunarch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


